### PR TITLE
perf: ensure compiled regex are compiled only once

### PR DESCRIPTION
**Pull request by pmiossec**
## Test methodology 

- manual

## Test environment(s) 

- Git Extensions 33.33.33
- Build 24e8371e13366c9de155109d4a319ef0c044f9a1 (Dirty)
- Git 2.38.0.windows.1 (recommended: 2.38.1 or later)
- Microsoft Windows NT 10.0.19042.0
- .NET 6.0.10
- DPI 96dpi (no scaling)

## Merge strategy

I agree that the maintainer squash merge this PR (if the commit message is clear).

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

### CODE REVIEW

**Review by gerhardol — COMMENTED** [review]
Static methods could be faster, maybe use compiled regex too.
https://learn.microsoft.com/en-us/dotnet/standard/base-types/best-practices#handle-object-instantiation-appropriately

**Review by RussKie — APPROVED** [review]

**Comment by RussKie** [comment]
.NET 7 introduces source generation for regex via attribute decorations.

**Comment by gerhardol** [comment]
> .NET 7 introduces source generation for regex via attribute decorations. 

I have considered reviewing all regex use for vNext. Most should use static methods I believe. Maybe wait for  net7 before looking in to that then.

**Comment by pmiossec** [comment]
> .NET 7 introduces source generation for regex via attribute decorations.

Yes looking forward to it but until that, a Compiled RegEx **must** be static otherwise it could degrade the performance instead of improving it...

File: GitCommands/Git/GitModule.cs [hunk 1]
```diff
@@ -3092,17 +3101,7 @@ public IReadOnlyList<string> GetMergedRemoteBranches()
...
         {
-            // Parse lines of format:
-            //
-            // 69a7c7a40230346778e7eebed809773a6bc45268 refs/heads/master
-            // 69a7c7a40230346778e7eebed809773a6bc45268 refs/remotes/origin/master
-            // 366dfba1abf6cb98d2934455713f3d190df2ba34 refs/tags/2.51
-            //
-            // Lines may also use \t as a column delimiter, such as output of "ls-remote --heads origin".
-
-            Regex regex = new(@"^(?<objectid>[0-9a-f]{40})[ \t](?<refname>.+)$", RegexOptions.Multiline | RegexOptions.Compiled);
```
Comment (gerhardol) [inline GitCommands/Git/GitModule.cs:3103]: Lines that start with sha should use TryParse() and trim whitespace for the rest, likely faster.
Maybe a separate PR.

**Review by gerhardol — APPROVED** [review]
There are some more Compiled Regex that are not static, but most can be ignored.
WordDictionary.cs:377 and maybe FileStatusList

Many of these should drop Compiled and use static methods (like Regex.Match(string, regex)
If you are executing in a loop, this is very efficient. I played around with this in the side panel stash PR, even if I did not need to use it. First call was about 1 ms for a quite simple Regex, then less than microseconds.

File: GitCommands/Git/GitModule.cs [hunk 2]
```diff
@@ -3337,6 +3336,8 @@ public GitBlame Blame(string? fileName, string from, Encoding encoding, string?
             }
         }
 
+        private static readonly Regex _headerRegex = new(@"^(?<objectid>[0-9a-f]{40}) (?<origlinenum>\d+) (?<finallinenum>\d+)", RegexOptions.Compiled);
```
Comment (RussKie) [inline GitCommands/Git/GitModule.cs:3339]: Please move this to the top of the file

Comment (mstv) [reply GitCommands/Git/GitModule.cs:3339]: @pmiossec, could you move it before leaving yet?

**Comment by pmiossec** [comment]
done.

**Review by mstv — APPROVED** [review]

**Pull request merged**